File: api/customer_organization.go
package api

import (
	"fmt"
	"slices"
	"time"

	"github.com/distr-sh/distr/internal/types"
	"github.com/distr-sh/distr/internal/validation"
	"github.com/google/uuid"
)

type CreateUpdateCustomerOrganizationRequest struct {
	Name     string                              `json:"name"`
	ImageID  *uuid.UUID                          `json:"imageId,omitempty"`
	Features []types.CustomerOrganizationFeature `json:"features,omitempty"`
}

func (r *CreateUpdateCustomerOrganizationRequest) Validate() error {
	if slices.Contains(r.Features, types.CustomerOrganizationFeatureAlerts) &&
		!slices.Contains(r.Features, types.CustomerOrganizationFeatureDeploymentTargets) {
		return validation.NewValidationFailedError(fmt.Sprintf("feature %v requires feature %v",
			types.CustomerOrganizationFeatureAlerts, types.CustomerOrganizationFeatureDeploymentTargets))
	}

	return nil
}

type CustomerOrganization struct {
	ID        uuid.UUID                           `json:"id"`
	CreatedAt time.Time                           `json:"createdAt"`
	Name      string                              `json:"name"`
	ImageID   *uuid.UUID                          `json:"imageId,omitempty"`
	ImageURL  *string                             `json:"imageUrl,omitempty"`
	Features  []types.CustomerOrganizationFeature `json:"features"`
}

type CustomerOrganizationWithUsage struct {
	CustomerOrganization
	UserCount             int64 `json:"userCount"`
	DeploymentTargetCount int64 `json:"deploymentTargetCount"`
}

File: sdk/js/src/types/customer-organization.ts
import {BaseModel} from './base';

export type CustomerOrganizationFeature = 'deployment_targets' | 'artifacts' | 'alerts' | 'support_bundles';

export interface CustomerOrganization extends Required<BaseModel> {
  name: string;
  imageId?: string;
  imageUrl?: string;
  features: CustomerOrganizationFeature[];
}

export interface CustomerOrganizationWithUsage extends CustomerOrganization {
  userCount: number;
  deploymentTargetCount: number;
}

export interface CreateUpdateCustomerOrganizationRequest {
  name: string;
  imageId?: string;
  features?: CustomerOrganizationFeature[];
}

export interface SidebarLink {
  id: string;
  createdAt: string;
  organizationId: string;
  customerOrganizationId?: string;
  name: string;
  link: string;
}

export interface CustomerOrganizationResponse extends CustomerOrganization {
  links: SidebarLink[];
}
